tests/ui/unwrap_or_else_default.rs

```diff
@@ -0,0 +1,41 @@
+// run-rustfix
+
+#![warn(clippy::unwrap_or_else_default)]
+#![allow(dead_code)]
+#![allow(clippy::unnecessary_wraps)]
+
+/// Checks implementation of the `OR_FUN_CALL` lint.
+fn or_fun_call() {
+    struct Foo;
+
+    impl Foo {
+        fn new() -> Foo {
+            Foo
+        }
+    }
+
+    enum Enum {
+        A(),
+    }
+
+    fn make<T, V>(_: V) -> T {
+        unimplemented!();
+    }
+
+    let with_enum = Some(Enum::A());
+    with_enum.unwrap_or_else(Enum::A);
+
+    let with_new = Some(vec![1]);
+    with_new.unwrap_or_else(Vec::new);
+
+    let with_err: Result<_, ()> = Ok(vec![1]);
+    with_err.unwrap_or_else(make);
+
+    let with_default_trait = Some(1);
+    with_default_trait.unwrap_or_else(Default::default);
+
+    let with_default_type = Some(1);
+    with_default_type.unwrap_or_else(u64::default);
```

Comment:
Applying the fixes from these in succession would get to the right place but I understand it would probably be better to be direct. Is the right thing to do to make this one see that and allow both lints to trigger, or should the redundant closure lint be suppressed?

Interestingly if you'd written `.unwrap_or(Default::default())` and (don't remember if this is required for this case) possibly the implementation of Default is nontrivial, it would tell you to use `.unwrap_or_default()` with yet another lint ;-)

Reply:
> Applying the fixes from these in succession would get to the right place

That's correct. The problem with two lint triggers is, that rustfix used

Reply:
> Applying the fixes from these in succession would get to the right place

That's correct. The problem with two lint triggers is, that rustfix can't automatically apply the lint suggestion and two messages for the same code snipped is also confusing to the user. We could use `clippy_utils::is_lint_allowed()` to check if both lints are allowed for this node and only suggest the rewrite with `.unwrap_or_default()` (This means suppressing `clippy::redundant_closure` for this case).

If you like, you can dive into this, otherwise we can wait for the feedback of someone else. I believe we have some other lints that work like this one currently does, meaning that it is most likely fine :)

Reply:
I think it was or_fun_call that did it.

Reply:
We have a few instances of inter-lint communication, but I would not block this PR on that. As long as the message is always valid and we get at the desired solution at some point, we're OK.

Reply:
I think successively applying the suggestions to get there is fine here. We can't write a `rustfix` test for it, but I think that should be ok.